src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/LanguageServiceHost.cs

```diff
@@ -132,7 +132,8 @@ protected override async Task InitializeCoreAsync(CancellationToken cancellation
                 target: DataflowBlockFactory.CreateActionBlock<IProjectVersionedValue<(ConfiguredProject ActiveConfiguredProject, ConfigurationSubscriptionSources Sources)>>(
                     async update => await ExecuteUnderLockAsync(cancellationToken => OnSlicesChanged(update, cancellationToken)),
```

Comment:
we can remove both async/await here to reduce extra async state machines.

also, I wonder whether you still need ExecuteUnderLockAsync here, because action block never processes two pieces of input at the same time, and the disposing code is to dispose links, and doesn't have any race condition between the two.  The async semaphore is essentially another queue to add more overhead to this.

Reply:
BTW, if we do have other code path to call this ExecuteUnderLockAsync, to follow the JTF rule, you must call this code inside a JTF.RunAsync:

async update => await _joinableTaskFactory.RunAsync(() => ExecuteUnderLockAsync(...)),

missing this JTF.RunAsync means that we will not join correctly for the async semaphore JTF chain, which can lead into deadlocks.

Reply:
The code in Workspace.cs looks good. I am not sure why we need retain the ExecuteUnderLock/AsyncSemaphore there as well. I cannot find any usage of WriteAsync methods. Those seem to be a part of old design, and can be removed?

Reply:
BTW, we should consider chaining the fault completion of the action block to abort the task completion source of _firstPrimaryWorkspaceSet. Basically, if we run into any exception inside OnSliceChanged, it will send NFE, but we don't want to hang the solution loading.  It just adds a safe net for this code.

Reply:
> I wonder whether you still need ExecuteUnderLockAsync here

Thanks, indeed `ExecuteUnderLockAsync` here seems redundant. Removed in b63e4434025bd4a9889024b6c806bac69607bc3a.

> The code in Workspace.cs looks good. I am not sure why we need retain the ExecuteUnderLock/AsyncSemaphore there as well. I cannot find any usage of WriteAsync methods. Those seem to be a part of old design, and can be removed?

They're used via `IWorkspaceWriter.WriteAsync` in TempPE, error list, properties providers and code model code. So we cannot remove them, and we must keep the `ExecuteUnderLockAsync` calls to prevent callers of `WriteAsync` overlapping with our own updates.

Does your advice around `_jtf.RunAsync(() => ExecuteUnderLockAsync(...` apply in `Workspace.cs` too then?

Reply:
Moved the removal of locking in `LanguageServiceHost` to a separate PR: https://github.com/dotnet/project-system/pull/8618